File: 2D-World-Simulator-Java/OPP-Java/src/op/JavaProject/Organisms/Animals/Human.java
package op.JavaProject.Organisms.Animals;

import op.JavaProject.Functionality.BoardPoint;
import op.JavaProject.Functionality.MoveDirection;
import op.JavaProject.Functionality.Species;
import op.JavaProject.Game.Animal;
import op.JavaProject.Game.Organism;
import op.JavaProject.Game.World;

import java.awt.*;
import java.io.Serializable;

import static op.JavaProject.Game.OrganismsStatics.*;

public class Human extends Animal implements Serializable {
    private MoveDirection direction = MoveDirection.NONE;
    private MagicalPotion ability = new MagicalPotion();

    //class of human special ability
    private class MagicalPotion implements Serializable {
        boolean activated;
        int coolDownTime;
        int newStrength;

        public MagicalPotion() {
            this.coolDownTime = 0;
            this.activated = false;
            this.newStrength = NEW_STRENGTH;
        }

    }

    public Human(World organismWorld, BoardPoint point) {
        super(organismWorld, point);
        this.strength = 5;
        this.initiative = 4;
        this.organismSign = "H";
        this.organismSpecie = Species.HUMAN;
        this.organismColor = new Color(254, 199, 203);
    }

    public void SetDirection(MoveDirection direction) {
        this.direction = direction;
    }

    //performing human move according to player given direction
    @Override
    public void Action() {
        ReduceCoolDownTime();
        if (ability.activated) {
            ReduceActivation();
        }

        int x = organismPos.GetX(), y = organismPos.GetY();
        BoardPoint newPoint = new BoardPoint();

        if (direction == MoveDirection.NONE) return;
        switch (direction) {
            case UP:
                newPoint = new BoardPoint(x, y - 1);
                break;
            case DOWN:
                newPoint = new BoardPoint(x, y + 1);
                break;
            case RIGHT:
                newPoint = new BoardPoint(x + 1, y);
                break;
            case LEFT:
                newPoint = new BoardPoint(x - 1, y);
                break;
            case DOWN_LEFT_EVEN:
                newPoint = new BoardPoint(x - 1, y + 1);
                break;
            case UP_LEFT_EVEN:
                newPoint = new BoardPoint(x - 1, y - 1);
                break;
            case DOWN_RIGHT_ODD:
                newPoint = new BoardPoint(x + 1, y + 1);
                break;
            case UP_RIGHT_ODD:
                newPoint = new BoardPoint(x + 1, y - 1);
                break;
            default:
                break;
        }

        Organism newPos = organismWorld.GetOrganismAtPos(newPoint);

        if (newPos != null) {
            Collision(newPos);
        } else {
            MoveToGivenField(newPoint);
        }
    }

    //activating human ability
    public void ActivateAbility() {

        if (ability.activated) {
            organismWorld.GetCommentator().AddComment("Human ability is already ACIVATED!");
            return;
        }

        if (ability.coolDownTime > 0) {
            organismWorld.GetCommentator().AddComment("Sorry, we cannot activate Human's ability. \n You need to wait " + ability.coolDownTime + " round/s.");
            return;
        }

        if (strength >= POWER_RANGE) {
            organismWorld.GetCommentator().AddComment("ERROR! Human already has power at least 10! ");
            return;
        }
        ability.activated = true;
        ability.newStrength = POWER_RANGE - strength;

        //strength rises to 10
        strength = GetStrength() + ability.newStrength;
        organismWorld.GetCommentator().AddComment("HUMAN ability ACTIVATED!");

    }

    //counting time left for activation
    private void ReduceActivation() {

        //changing human's strength
        if (ability.newStrength > 0) {
            ability.newStrength--;
            strength--;
        } else {
            //we need to switch off ability power
            //strength is == 5, the initial state -> so we do not have to change it
            organismWorld.GetCommentator().AddComment("HUMAN ability DISACTIVATED!");
            ability.activated = false;
            ability.coolDownTime = COOL_DOWN_TIME;
        }
    }

    //reducing cooldown time after activation of ability
    private void ReduceCoolDownTime() {
        if (ability.coolDownTime > 0) ability.coolDownTime--;
    }

    public MoveDirection GetDirection() {
        return this.direction;
    }

}

File: 2D-World-Simulator-Python/OPP-Python/Organisms/Animals/Human.py
from Game.Animal import Animal
from Game.OrganismStatics import *


class MagicalPotion:
    def __init__(self):
        self.coolDownTime = 0
        self.activated = False
        self.new_strength = NEW_STRENGTH


class Human(Animal):

    def __init__(self, organism_world, point):
        super().__init__(organism_world, point, 5, 4, "H", OrganismSpecie.HUMAN, "#fec7cb")
        self.direction = NONE
        self.direction_point = None
        self.ability = MagicalPotion()

    # performing human move according to player given direction
    def action(self):
        self.reduce_cooldown_time()
        if self.ability.activated:
            self.reduce_activation()

        if self.direction == NONE:
            return

        new_pos = self.organism_world.get_organism_at_pos(self.direction_point)

        # checking there is other organism
        if new_pos is not None:
            super().collision(new_pos)
        else:
            self.move_to_given_field(self.direction_point.x, self.direction_point.y)

        self.direction = NONE
        self.direction_point = None

    # activate human's ability
    def activate_ability(self):
        if self.ability.activated:
            self.organism_world.get_world_window().add_comment(
                "Human ability is already ACTIVATED!")
            return

        if self.ability.coolDownTime > 0:
            self.organism_world.get_world_window().add_comment("Sorry, we cannot activate Human's ability." + "\n" +
                                                               " You need to wait " +
                                                               str(self.ability.coolDownTime) + " round/s.")
            return

        if self.strength >= POWER_RANGE:
            self.organism_world.get_world_window().add_comment("ERROR! Human already has power at least 10! ")
            return

        self.ability.activated = True
        self.ability.new_strength = POWER_RANGE - self.strength

        # strength rises to 10
        self.strength = POWER_RANGE

        self.organism_world.get_world_window().add_comment("HUMAN ability ACTIVATED!")

    # counting time left for activation
    def reduce_activation(self):
        # changing human's strength
        if self.ability.new_strength > 0:
            self.ability.new_strength -= 1
            self.strength -= 1
        else:
            # we need to switch off ability power
            # strength at its initial state -> so we do not have to change it
            self.organism_world.get_world_window().add_comment("HUMAN ability DISACTIVATED!")
            self.ability.activated = False
            self.ability.coolDownTime = COOL_DOWN_TIME

    # reducing cooldown time after activation of ability
    def reduce_cooldown_time(self):
        if self.ability.coolDownTime > 0:
            self.ability.coolDownTime -= 1

    def set_direction(self, new_direction):
        self.direction = new_direction

    def get_direction(self):
        return self.direction

    def set_direction_point(self, point):
        self.direction_point = point
